File: java/say/src/main/java/Say.java
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class Say {

    private static final List<String> small = List.of(
            "zero",
            "one",
            "two",
            "three",
            "four",
            "five",
            "six",
            "seven",
            "eight",
            "nine",
            "ten",
            "eleven",
            "twelve",
            "thirteen",
            "fourteen",
            "fifteen",
            "sixteen",
            "seventeen",
            "eighteen",
            "nineteen"
    );
    private static final List<String> tens = List.of("", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety");
    private static final List<String> scales = List.of("", "thousand", "million", "billion");

    public String say(long number) {
        if (number < 0 || number > 999_999_999_999L) {
            throw new IllegalArgumentException("");
        }

        if (number == 0) {
            return "zero";
        }

        var num = number;
        var scaleIndex = 0;
        final var parts = new ArrayList<String>();

        while (num > 0) {
            final var chunk = (int) (num % 1000);

            if (chunk != 0) {
                final var chunkText = sayUnder1000(chunk);
                final var scale = scales.get(scaleIndex);

                parts.add(scale.isEmpty() ? chunkText : "%s %s".formatted(chunkText, scale));
            }

            num /= 1000;
            scaleIndex++;
        }

        return String.join(" ", parts.reversed());
    }

    private String sayUnder1000(int num) {
        final var parts = new ArrayList<String>();

        final var hundreds = num / 100;
        final var remainder = num % 100;

        if (hundreds > 0) {
            parts.add(small.get(hundreds) + " hundred");
        }

        if (remainder > 0) {
            if (remainder < 20) {
                parts.add(small.get(remainder));
            } else if (remainder % 10 == 0) {
                parts.add(tens.get(remainder / 10));
            } else {
                parts.add("%s-%s".formatted(tens.get(remainder / 10), small.get(remainder % 10)));
            }
        }

        return String.join(" ", parts);
    }
}

File: kotlin/say/src/main/kotlin/Say.kt
class NumberSpeller {

    private val small = listOf(
        "zero",
        "one",
        "two",
        "three",
        "four",
        "five",
        "six",
        "seven",
        "eight",
        "nine",
        "ten",
        "eleven",
        "twelve",
        "thirteen",
        "fourteen",
        "fifteen",
        "sixteen",
        "seventeen",
        "eighteen",
        "nineteen",
    )
    private val tens = listOf("", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety")
    private val scales = listOf("", "thousand", "million", "billion")

    fun say(input: Long): String {

        require(input in 0..999_999_999_999) {
            "Number out of range"
        }

        if (input == 0L) return "zero"

        var number = input
        var scaleIndex = 0
        val parts = mutableListOf<String>()

        while (number > 0) {
            val chunk = (number % 1000).toInt()

            if (chunk != 0) {
                val chunkText = sayUnder1000(chunk)
                val scale = scales[scaleIndex]

                parts += if (scale.isEmpty()) {
                    chunkText
                } else {
                    "$chunkText $scale"
                }
            }

            number /= 1000
            scaleIndex++
        }

        return parts.asReversed().joinToString(" ")
    }

    private fun sayUnder1000(number: Int): String {
        val parts = mutableListOf<String>()

        val hundreds = number / 100
        val remainder = number % 100

        if (hundreds > 0) {
            parts += "${small[hundreds]} hundred"
        }

        if (remainder > 0) {
            val text = when {
                remainder < 20 -> small[remainder]

                remainder % 10 == 0 -> {
                    tens[remainder / 10]
                }

                else -> {
                    "${tens[remainder / 10]}-${small[remainder % 10]}"
                }
            }

            parts += text
        }

        return parts.joinToString(" ")
    }
}
